main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolverInterface.scala

```diff
@@ -192,6 +205,59 @@ object ConstraintSolverInterface {
     case JvmMember.JvmStaticMethod(clazz, name, tpes) => TypeError.StaticMethodNotFound(clazz, name, tpes.map(subst.apply), renv, loc)
   }
 
+  /**
+    * Creates appropriate errors for the mismatched record types.
+    */
+  private def mkErrorsFromRecords(tpe1: Type, tpe2: Type, renv: RigidityEnv, loc: SourceLocation)(implicit flix: Flix): List[TypeError] = {
+    // TODO: Open records
+    (tpe1, tpe2) match {
+      case (Type.Var(_, _), _) | (_, Type.Var(_, _)) => List(mkMismatchedTypesOrEffects(tpe1, tpe2, tpe1, tpe2, renv, loc))
+      case _ =>
+        val tpe1Labels = getRecordLabels(tpe1, List.empty[(Name.Label, Type)])
+        val tpe2Labels = getRecordLabels(tpe2, List.empty[(Name.Label, Type)])
+        // Extra label only.
+        if (tpe1Labels.isEmpty && !tpe2Labels.isEmpty) {
+          tpe2Labels.toList.map {
+            case (label, labelTpe) =>
+              TypeError.ExtraLabel(label, labelTpe, tpe2, renv, loc)
+          }
+          // Undefined label only.
+        } else if (tpe2Labels.unzip._1.toSet.subsetOf(tpe1Labels.unzip._1.toSet)) {
+          tpe1Labels.toList.flatMap {
+            case (label, labelTpe) => {
+              if (!tpe2Labels.unzip._1.contains(label)) {
+                List(TypeError.UndefinedLabel(label, labelTpe, tpe2, renv, loc))
+              } else Nil
+            }
+          }
+        } else if (tpe2Labels.isEmpty && !tpe1Labels.isEmpty) {
+          tpe1Labels.toList.map {
+            case (label, labelTpe) =>
+              TypeError.UndefinedLabel(label, labelTpe, tpe2, renv, loc)
+          }
+          // Both undefined and extra labels.
+        } else {
```

Comment:
I think we only need this else-branch, right? Because this covers the empty cases.

Reply:
I'm still confused.

I would expect it to be essentially:
- `extraLabels = tpe1Labels - tpe2Labels`
- `missingLabels = tpe2Labels - tpe1Labels`

Which is what you have below. All the empty/subset checking is handled by this, no?